rest_specification/restful_service_specification.md

```diff
@@ -0,0 +1,240 @@
+|----------------|------------------------|-------------------------------|--------------|----------------|
+| `POST`         | Create new resource(s) | -                             | JSON array   | -              |
+| `PUT`          | Replace resource(s)    | -                             | JSON array   | -              |
+| `PATCH`        | Update Resource(s)     | -                             | JSON array   | -              |
+| `GET`          | List resource(s)       | `fields`,`page`,`size`,`sort` | -            | paginated list |
+| `DELETE`       | Delete ALL resources   | -                             | -            | -              |
+
+Endpoints may not support all methods (i.e. a service may provide read-only
+access to a resource, or a service may not support certain operations for
+performance or other considerations; for example, `GET` may not be supported
+for `user` if listing all users is disallowed for performance reasons.
+
+## Operations Against Specific Resources
+Specific resources will be addressed through the base URI plus the identifier
+for that resource.
+
+    /user/{id}
+
+The following operations may be supported:
+
+| HTTP Operation | Description                   | Query Parameters | Request Body | Response Body |
+|----------------|-------------------------------|------------------|--------------|---------------|
+| `GET`          | Retrieve the resource         | `fields`         | -            | JSON object   |
+| `PUT`          | Replace the existing resource | -                | JSON object  | -             |
+| `PATCH`        | Update an existing resource   | -                | JSON object  | -             |
+| `DELETE`       | Delete the resource           | -                | -            | -             |
+
+Clients should use the `fields` parameter for better performance.
+
+Implementors should consider breaking out field represented by a JSON array
+into sub-entities if there are performance concerns around its size. This
+consideration should happen regardless of how the data is stored.
+
+## Pagination
+
+Paging shall be implemented via `page` and `size` query parameters. If none are
+provided, the default page size shall be used.
+
+The default page-size is implementation-specific.
+
+| Parameter | Description               |
+|-----------|---------------------------|
+| `page`    | page index, starting at 0 |
+| `size`    | page size                 |
+
+## Sorting
+
+Sorting shall be provided via the `sort` query parameter. `sort` may be
+specified more than once in order to sort by more than one field. `sort` values
+shall be of the form `sort={field}(,{direction})` where `direction` is `asc` or
+`desc`.
+
+### Responses
+
+The response shall be a JSON object formatted according to Spring Data REST
+conventions.Specifically, the list is returned as a pluralized field name
+under `_embedded`, and a `page` object supplies paging info. Example:
+
+```json
+GET /user
+
+{
+  "_embedded": {
```

Comment:
Two questions WRT responses:

__1: What about error responses for CRUD operations?__
There are some standards here:
https://github.com/blongden/vnd.error
https://datatracker.ietf.org/doc/rfc7807/?include_text=1

We could also consider just using JSend error responses that we use for RPC calls below.

__2: Why have different responses for CRUD operations vs additional functions?__
I'd like to see us accommodate both cases with one standard. Ideally, all successes look consistent and all errors look consistent. To me they are not different enough to warrant having to understand how to parse different responses. Again, maybe not so bad for those of us comfortable with JSON, HTTP, HAL, etc, but trying to empathize with other users. And still, it would be nice for developer sanity, too, I think (imagine trying to overlay multiple frameworks / libraries for different standards vs just using one for one standard). Plus, it shouldn't be so hard to consolidate, they are effectively just responses with data either way. This is consistent with most examples I've seen in the wild, too (they all use the same response envelope / metadata).

So, why not use HAL for additional function responses, too? Or JSend for both. I don't really care what standard, just that we're not arbitrarily inconsistent. What do you guys think?